### DISCUSSION TITLE
GetOrCreateTool throws "The name of the function does not conform to naming standards: ^[a-zA-Z0-9_-]{1,64}$'" exception

### DISCUSSION THREAD

**Discussion by MartinKlepal** [RageAgainstThePixel/OpenAI-DotNet#271] [Q&A]
what ever I try I always get the exception:

for example, I define a class
```csharp
        static public class TestObject
        {
            static public string Test()
            {
                 return "text";         
            }
        };
```
and then
```csharp
Tool t = Tool.GetOrCreateTool( typeof( TestObject), nameof( TestObject.Test));
```
this throws the "The name of the function does not conform to naming standards: ^[a-zA-Z0-9_-]{1,64}$'" exception

any advice appreciated.

**Comment by StephenHodgson** [comment]
Enable debug mode, and print your results here

**Reply by MartinKlepal** [reply]
this is the detail for the exception from Visual Studio. (I guess this is not what you mean)

System.ArgumentException
  HResult=0x80070057
  Message=The name of the function does not conform to naming standards: ^[a-zA-Z0-9_-]{1,64}$
  Source=OpenAI-DotNet
  StackTrace:
   at OpenAI.Function..ctor(String name, String description, MethodInfo method, Object instance)
   at OpenAI.Tool.GetOrCreateTool(Type type, String methodName, String description)
   at ...

I can see EnableDebug (if that is what you are referring to) is defined in OpenAIBaseEndpoint class but not in Tool class
I do not know how to enable debug mode for the exception above in the Tool class, be honest....

**Reply by StephenHodgson** [reply]
I'll add the name that it's trying to parse in my next update. 

In the meantime try doing a different name.

**Reply by MartinKlepal** [reply]
I tried all different tweaks and names already with the same result....

**Reply by StephenHodgson** [reply]
Take a look at one of my unit tests, and try that

**Reply by MartinKlepal** [reply]
I copied pasted from the unit test following WeatherService class: 

        internal static class WeatherService
        {
            internal enum WeatherUnit
            {
                Celsius,
                Fahrenheit
            }

            [Function("Get the current weather in a given location.")]
            public static async Task<string> GetCurrentWeatherAsync(
               [FunctionParameter("The location the user is currently in.")] string location,
               [FunctionParameter("The units the user has requested temperature in. Typically this is based on the users location.")] WeatherUnit unit)
            {
                var temp = new Random().Next(-10, 40);

                temp = unit switch
                {
                    WeatherUnit.Fahrenheit => CelsiusToFahrenheit(temp),
                    _ => temp
                };

                return await Task.FromResult($"The current weather in {location} is {temp}\u00b0 {unit}");
            }

            public static int CelsiusToFahrenheit(int celsius) => (celsius * 9 / 5) + 32;
        }

and called  

Tool.GetOrCreateTool(typeof(WeatherService), nameof(WeatherService.GetCurrentWeatherAsync));

which throws the same exception :(

**Reply by StephenHodgson** [reply]
Sounds like you are somehow changing the JSON serialization settings in your project

**Reply by MartinKlepal** [reply]
the testing project is in .net 7.0 with WPF, it includes newtonsoft.json, but besides that I do not change json serialisation setting in the project

**Reply by StephenHodgson** [reply]
This project uses the system.text.json.

Unsure if that might be causing your problem. Try removing the Newtonsoft Lib

**Reply by MartinKlepal** [reply]
created a fresh new WPF project without anything and without newtonsoft.json but the same exception still
